core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala

```diff
@@ -55,7 +55,8 @@ case class SparkListenerTaskGettingResult(taskInfo: TaskInfo) extends SparkListe
 @DeveloperApi
 case class SparkListenerSpeculativeTaskSubmitted(
```

Comment:
If we need to ensure backward compatibility, can we try to change `SparkListenerSpeculativeTaskSubmitted` like `SparkListenerJobStart`?

https://github.com/apache/spark/blob/3fc8a90267312fdfc42b7e6a16ee69f2507ef56b/core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala#L73-L83

Reply:
Fix like this:
<img width="856" alt="image" src="https://user-images.githubusercontent.com/20420642/205205060-071827bf-bba2-47bc-b98d-7c778e00820d.png">
<img width="878" alt="image" src="https://user-images.githubusercontent.com/20420642/205205206-fecb2fe3-bbc7-49e0-bf60-af5cc3a09f98.png">

Reply:
If you want `taskIndex` to be a `val`, maybe we can consider using a size 1 `Array` as an container, and add two function to update and get the Array(0) as `taskIndex`

Reply:
Interesting option, but I would not prefer it to be modifiable by users - given it is a developer api.
One option would be to do something like:
```

// modify json protocol, etc to 
private[spark] var _taskIndex: Int = -1

def taskIndex  = _taskIndex

```

+CC @jiangxb1987 who had worked on this in the past.
Thoughts @Ngone51

Reply:
> Interesting option, but I would not prefer it to be modifiable by users - given it is a developer api. One option would be to do something like:
> 
> ```
> 
> // modify json protocol, etc to 
> private[spark] var _taskIndex: Int = -1
> 
> def taskIndex: Int = _taskIndex
> ```
> 
> +CC @jiangxb1987 who had worked on this in the past. Thoughts @Ngone51

I think this is ok

Reply:
+1 to @mridulm 's option.

Reply:
+1 on a spark private variable.

Reply:
Since we are making an API change, given SPARK-37831, we should add `partitionId` as well.

Reply:
Since `partitionId` is never used in this PR, is it only added for possible future use? 
If so, I'd prefer to add this in another pr to solve problem given [SPARK-37831](https://issues.apache.org/jira/browse/SPARK-37831)~ @mridulm

Reply:
Is it a lot of change to add support for it ?
I would want to keep both of them in sync, and prevent divergence - given the issues we have already faced with that.

Reply:
`Is it a lot of change to add support for it ?`
If you just want to add `partitionId` as a field of taskEvent, the change will not be too much.
`I would want to keep both of them in sync, and prevent divergence`

Reply:
`Is it a lot of change to add support for it ?`
If you just want to add partitionId as a field of taskEvent, the change will not be too much. The question is why do we need `partitionId` in `ExecutorAllocationManager`?
`I would want to keep both of them in sync, and prevent divergence`
But if you want to keep `partitionId` and `taskIndex` in sync or use `partitionId` instead of `taskIndex`,  the changes will be huge, and I don’t think it’s appropriate to make changes here~

Reply:
Got it, I'll make a commit to try and fix this soon